src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1384,21 +1384,30 @@ AddNodeMetadata(char *nodeName, int32 nodePort,
 	*nodeAlreadyExists = false;
 
 	/*
-	 * Take an exclusive lock on pg_dist_node to serialize node changes.
-	 * We may want to relax or have more fine-grained locking in the future
-	 * to allow users to add multiple nodes concurrently.
+	 * Prevent / wait for concurrent modification before checking whether
+	 * the worker already exists in pg_dist_node.
 	 */
-	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
+	LockRelationOid(DistNodeRelationId(), RowShareLock);
```

Comment:
`DeleteNodeRow` seems to acquire `RowExclusiveLock`, so they do not conflict anymore. I guess there is a small probability of `workerNode ` being freed if it is deleted concurrently?

So, maybe `DeleteNodeRow` should acquire `ExclusiveLock `?

Reply:
This seems to have changed from AccessExclusive to RowExclusive on this PR/commit: https://github.com/citusdata/citus/pull/1476/commits/bd369d971d11abdf47e77a18c0a2c52079e03258

I'm still looking the side effects of changing the lock to `ExclusiveLock ` in `DeleteNodeRow `

Reply:
hmz, good catch. I guessing the exclusive lock in add node is mainly to avoid missing a worker during operations like type propagation, while there is no such risk when concurrently removing a worker.

I think an ExclusiveLock might work, but I'll think about it a little more. Alternative could be to recheck after getting the exclusive lock.

Reply:
Added a second check.

Concurrent remove node is not an issue, because we either see a worker node during the first node (because remove has committed), and return, or we don't (because remove has committed) and we add it. 

However, 2 concurrent add nodes could compete for the exclusive lock and we expect the second one to be idempotent. Hence, the second check is needed.